### PR TITLE
Add eslint and prettier and fix lint issues

### LINKED ISSUE #28: Adopt Prettier / Eslint?
What do you think about using [prettier](https://github.com/prettier/prettier) @alidcastano ?

Seems a good way of avoiding debates about code style.

### PR DESCRIPTION
**Pull request by alexparish**
ESLint 😍
Prettier with husky + lint-staged 😍

_Notes:_

- Used airbnb config
- Ignoring `src/bundle/templates` as there are some funky placeholders in there
- Modified `react/jsx-filename-extension` rule to allow JSX in files with `.js` extension
- Renamed `bin/rogue` to `bin/rogue.js` so it's picked up by the linter
- Added `lint` script to package.json for future use in CI pipeline

Closes #28 

Let me know if you have any questions! More than happy to make updates.

### CODE REVIEW

**Comment by peter-mouland** [comment]
lovely stuff !

2 questions though...
 * as`husky` v1 is a breaking change, is it worth getting the `.rc-12` in to stay ahead of the curve, or you guys prefer to fix as and when the _v1_ is released?
 * I've not used `lint-staged` before, not sure how I completely get it.  is it committing for you, or just adding files?
   * if committing, not sure how I feel about that, i guess i'd get use to it or cancel the commit if i wanted to squish?
   * if just adding files, does that mean the tests will pass locally, but the changes wont get pushed so may fail on CI?

**Comment by alexparish** [comment]
`husky`: Good shout. l will update to the release candidate (unless I hear any objections). I particularly like that hooks have been moved to their own `husky` property in package.json. 

`lint-staged`: The changes prettier makes are added to the existing commit. Never had any problems with that on previous projects but I imagine there are icky edge cases. I need to have a further think about the CI case. The approach is documented here: https://prettier.io/docs/en/precommit.html

**Comment by peter-mouland** [comment]
cool - I'm probably a bit old-skool and do like to review each commit before it's made, but i guess prettier is simply spacing (not as complex as eslint) so might be safe without a review.  

...on that es eslint not going to be run on a `pre-commit` hook also?

**Comment by alexparish** [comment]
> is eslint not going to be run on a `pre-commit` hook also?

I added https://github.com/prettier/eslint-plugin-prettier so that eslint runs the prettier rules. The docs for that plugin say "if you run eslint with the --fix flag, your code will be formatted according to prettier style"

So if I change `prettier --write` to `eslint --fix` in the `lint-staged` section we should be golden.

**Comment by alexparish** [comment]
Have updated the pull request with the following:

- Use `husky@1.0.0-rc.12`
- Run eslint instead of prettier in pre-commit hook

File: examples/hello-world/package.json [hunk 1]
```diff
@@ -10,6 +10,9 @@
   "author": "",
   "license": "ISC",
   "dependencies": {
+    "react": "^16.4.1",
+    "react-dom": "^16.4.1",
```
Comment (alidcastano) [inline examples/hello-world/package.json]: Should we be adding these as dependencies?

I followed Nextjs on this which makes the user instal it themselves

**Comment by alidcastano** [comment]
going to make comments on PR

Also, I don't mean to bikeshed, but:
- I generally prefer single quotes over double quotes
- I strongly dislike semicolons 

Not sure what the react community generally prefers, and I forget which eslint config I used to use since I've lately been using typescript so sorry for not bringing this up earlier

File: package.json [hunk 2]
```diff
@@ -1,56 +1,80 @@
...
+    "parcel-bundler": "^1.9.2",
+    "react": "^16.4.1",
+    "react-apollo": "^2.1.8",
+    "react-dev-utils": "^5.0.1",
+    "react-helmet": "^5.2.0",
+    "react-redux": "^5.0.7",
+    "react-router": "^4.3.1",
+    "react-tree-walker": "^4.1.0",
+    "serialize-javascript": "^1.5.0",
+    "styled-components": "^3.3.3"
```
Comment (alidcastano) [inline package.json]: some devDependcies changed to dependencies here too

e.g. react-redux and styled-components - how come? 

They're optional for user to use, we only install them so we can run hoc ourselves if we need to for testing

**Comment by alidcastano** [comment]
@alexparish I'm working on switching from express to connect / programmatic usage right now, which is a decently sized refactor anyway so I can try to consolidate these changes based on your code or if you prefer I can wait for us to merge this PR & then work off that, so lmk 

and thanks for getting this going

File: package.json (hunk 2 above)
Comment (alexparish) [reply package.json]: Again, these were moved due to lint errors. For example:

`error  'react-redux' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies`

**Comment by alexparish** [comment]
@alidcastano I felt the same as you about semicolons (I used to use `standard` for linting) and was passionately against them but I have changed my mind for various reasons. If it's just the case that you don't like typing them at the end of every line... you can install an extension for your text editor of choice to run prettier when you save a file and it'll add semicolons for you as part of the prettification.

From what I've seen the community seems to be converging on prettier + airbnb rules (or similar) which include semicolons. Ultimately you're the maintainer so it's your decision and I understand the arguments for both.

I don't mind how you want to proceed with this pull request. Happy to rebase it once you've made your changes and equally happy for you to reimplement these changes yourself in another pull request. I would strongly recommend using prettier though, in conjunction with a git hook.

File: examples/hello-world/package.json (hunk 1 above)
Comment (alidcastano) [reply examples/hello-world/package.json]: Hmm I wonder if we should just install `react react-dom react-router` for users or if it's better to keep them as peer dependencies

nextjs has react as a peer dependency, whereas nuxtjs has vue as a dependency


either way this is what nextjs uses:

```
	  "standard": {
    "parser": "babel-eslint",
    "ignore": [
      "**/node_modules/**",
      "**/examples/**/lib/**"
    ]
  }
```

and btw they use single quotes and no semi colons so lets us whatever they're using!

File: package.json (hunk 2 above)
Comment (alidcastano) [reply package.json]: in this case for sure I'd rather ignore linter errors than to add unnecessary dependencies

**Comment by alidcastano** [comment]
yeah, as I wrote in snippet above, let's stick with single quotes and no semicolons. Nextjs does this too so we can copy their config them on this

I'll try to get my changes in soon, so you can rebase PR; thanks

**Comment by peter-mouland** [comment]
I'll just add my 2cents, even though it sounds like your set, but there are a number of good reasons _for_ semi colons.  like : 
 * not having to rely on ASI (semi colons _will_ be inserted, so best we do it first so we can check)
 *  being able start lines with any character : 
    * i.e using semi colons we can start using `['my, 'items'].includes(myVar)`
    *  without semi colons it has to be `;['my,'items'].includes(myVar)`
 * without them leaves room for ambiguity and misunderstanding

I think removing any pause/cause for thought it best, especially in collaborative or OS projects.

The ambiguity argument for me is the same as why I have multi-line braces for conditionals.

e.g. without we could have : 
```
if (condition)
   do_something()
   do_something_else()
```
now, go to review that code and did the dev mean : 
```
if (confidtion) {
   do_something()
}
do_something_else()
```
or did they mean
```
if (condition) {
   do_something()
   do_something_else()
}
```

I think ambiguity slows down the dev later as they'll have to re-read the whole snippet/file to truly understand what was intended. 

Thinking in terms of minifying, is the reviewer/dev always going to remember to start lines with a semi-colon when needed or will bugs creep in.  The following : 

```
var a = myObj.prop
['my','bug'].map()
```
gets minified to

```
var a = myObj.prop.my.bug.map()
```
where as we actually intend it to be

```
var a=myObj.prop;["my","bug"].map();
```

best to remove any ambiguity at the source.

**Comment by alexparish** [comment]
Seems like your mind is made up!

If I was to re-implement these changes on the current master branch I'd like to be clear on what standard prettier rules you'd like to be changed.

Is it just disabling semicolons and using single quotes instead of double quotes?

**Comment by alexparish** [comment]
I think that would come under the scope of this eslint rule: https://eslint.org/docs/rules/curly

Could you confirm that you'd like the `multi-or-nest` rule?

**Comment by alidcastano** [comment]
Yeah that works!

**Comment by alidcastano** [comment]
ok so bad news, good news 

bad news: had to add a build step (#46) ... and figured I'd refactor `roguejs/app` to typescript given that. So I'm sure that complicates this PR 

good news: API is beautifully simple now

**Comment by peter-mouland** [comment]
lol - Iet's hope most of them were fixed with `--fix`!  have fun @alexparish  👍

**Comment by alexparish** [comment]
Have been busy and now on holiday for a week so if anyone fancies having a go at this (start over from master branch) then please go ahead.

**Pull request closed without merging**